src/rpc/api/modules/validator.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  Attestation, AttestationData, BeaconBlock, bytes32, Deposit, Shard, Slot, Eth1Data, uint64,
+  Fork, SyncingStatus, ValidatorDuty, bytes48, bytes, IndexedAttestation, number64
+} from "../../../types";
+import {DB} from "../../../db";
+import {BeaconChain} from "../../../chain";
+import {OpPool} from "../../../opPool";
+
+import {IValidatorApi} from "../interfaces";
+
+export class ValidatorApi implements IValidatorApi {
+  private chain: BeaconChain;
+  private db: DB;
+  private opPool: OpPool;
+
+  public constructor(opts, {chain, db, opPool}) {
+    this.chain = chain;
+    this.db = db;
+    this.opPool = opPool;
+  }
+
+  public async getClientVersion(): Promise<bytes32> {
+    return Buffer.alloc(32);
+  }
+
+  public async getFork(): Promise<{fork: Fork; chainId: number64}> {
+    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
+    return {} as {fork: Fork; chainId: number64};
+  }
+
+  public async getGenesisTime(): Promise<number64> {
+    return await this.chain.genesisTime;
+  }
+
+  public async getSyncingStatus(): Promise<boolean | SyncingStatus> {
+    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
+    return {} as boolean | SyncingStatus;
+  }
+
+  public async getDuties(validatorPubkeys: bytes48[]): Promise<{currentVersion: Fork; validatorDuties: ValidatorDuty[]}> {
+    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
+    return {} as {currentVersion: Fork; validatorDuties: ValidatorDuty[]};
+  }
+
+  public async produceBlock(slot: Slot, randaoReveal: bytes): Promise<BeaconBlock> {
```

Comment:
I think the validator should be able assemble the block themselves, to the extent possible.

This function minimizes the number of requests, but it obscures a lot of validator logic that ends up offloaded to the beacon chain.

Maybe I would feel better about this if we exposed both a high-level and low-level interface. The high-level interface would use the low-level interface.

eg something like:

```typescript
class ValidatorApi {
  ...
  // low level api
  public async getProposerSlashings(): Promise<ProposerSlashing[]> {}
  public async getAttesterSlashings(): Promise<AttesterSlashing[]> {}
  public async getDeposits(): Promise<Deposit[]> {}
  public async getVoluntaryExits(): Promise<VoluntaryExit[]> {}
  public async getTransactions(): Promise<Transaction[]> {}
  public async getBlockBody(): Promise<BlockBody> {
    return {
      proposerSlashings: await this.getProposerSlashings(),
      ...other operations
    };
  }
  public async computeStateRoot(block: BeaconBlock): bytes32 {}
  ...etc

  // high level api
  public async produceBlock(slot: Slot): Promise<BeaconBlock> {
    const body = await this.getBlockBody();
    ...
    block.stateRoot =  await this.computeStateRoot(block);
    return block;
  }
```
This would give us more flexibility in the future by exposing the underlying logic and letting unknown validator implementations potentially do more. (eg: more easily/flexibly fetch/merge data from multiple beacon nodes)

Reply:
A lot of this endpoints is not strictly related to validator. How about we separate them into ValidatorApi and BeaconApi exposed with namespacing.I can imagine that BeaconApi would be required by block explorers and tools. It's already supported in noice-json-rpc library. Example we expose BeaconApi with namespace `beacon_*` and ValidatorApi with namespace `validator_*`. I can quickly add it to this PR if we come to a conclusion :)

Reply:
Yeah exactly, I suggested that we should probably seperate them, hence why I changed up the structure of the rpc section in general.

So basically, my idea would be json-rpc.ts (or whatever it is) we can add a field to the api class called "namespace" and then when combining them we can namespace that way.

That being said, @wemeetagain I'm not sure we need to expose all these methods because I want this to be explicitly for the validators.